File: src/main/golang/com/majun/leetcode/CQueue.go
package leetcode

import "container/list"

type CQueue struct {
	stack, reverseStack *list.List
}

func Constructor() CQueue {
	return CQueue{
		stack:        list.New(),
		reverseStack: list.New(),
	}
}

func (this *CQueue) AppendTail(value int) {
	this.stack.PushBack(value)
}

func (this *CQueue) DeleteHead() int {
	if this.reverseStack.Len() == 0 {
		for this.stack.Len() > 0 {
			this.reverseStack.PushBack(this.stack.Remove(this.stack.Back()))
		}
	}
	if this.reverseStack.Len() != 0 {
		head := this.reverseStack.Back()
		this.reverseStack.Remove(head)
		return head.Value.(int)
	}
	return -1
}

File: src/main/java/com/majun/leetcode/CQueue.java
package com.majun.leetcode;

import java.util.ArrayDeque;
import java.util.Deque;

//测试gitcheck
/**
 * @author majun 2021-11-27
 * 用两个栈实现一个队列。队列的声明如下，请实现它的两个函数 appendTail 和 deleteHead ，分别完成在队列尾部插入整数和在队列头部删除整数的功能。(若队列中没有元素，deleteHead操作返回 -1 )
 * 来源：力扣（LeetCode）
 * 链接：https://leetcode-cn.com/problems/yong-liang-ge-zhan-shi-xian-dui-lie-lcof
 */
class CQueue {

    Deque<Integer> stack = new ArrayDeque<>();
    Deque<Integer> reverseStack = new ArrayDeque<>();


    public void appendTail(int value) {
        stack.push(value);
    }

    public int deleteHead() {
        if (reverseStack.isEmpty() && stack.isEmpty()) {
            return -1;
        }
        if (!reverseStack.isEmpty()) {
            return reverseStack.pop();
        }
        int size = stack.size();
        for (int i = 0; i < size; i++) {
            Integer pop = stack.pop();
            reverseStack.push(pop);
        }
        return reverseStack.pop();
    }
}